### ISSUE TITLE
Time zone incorrect

### ISSUE THREAD

**Issue by schristm** [basecamp/local_time#41]
Hello,

For some reason, local_time is showing time stamps in the future by 4 hours. My application.rb does not change the default time zone from UTC, however there is a before_action that sets the Time.zone to the current user's time zone, which in my case is "Eastern Time (US & Canada)".

Anywhere I render a date without using local_time, the time zone conversion is accurate, however when I substitute with local_time, I'm jumped ahead by 4 hours. Can you help point me in the right direction?

**Comment by javan** [comment]
Do you have local_time included in your JavaScript bundle (`//= require local_time`)? If that's not the issue, paste an example of how you're using the the local time helpers in your views and maybe that'll reveal the problem.

**Comment by schristm** [comment]
Thanks for getting back to me. I am including the JavaScript bundle, which I've verified is loading properly and updates the text in the time tag. Here's how I'm using the time helper:

> <%= local_relative_time @payment.created_at, type: 'weekday-or-date' %>

**Comment by schristm** [comment]
For reference, if I use the rails-timeago plugin, my time tag is proper:

> <time datetime="2015-07-02T02:30:56-04:00" title="Thu, 02 Jul 2015 02:30:56 -0400" data-time-ago="2015-07-02T02:30:56-04:00">about 12 hours ago</time>

When using local_time, my time tag is 4 hours ahead:

> <time datetime="2015-07-02T06:30:56Z" data-local="weekday-or-date" title="July 2, 2015 at 6:30am EDT">Today</time>

**Comment by javan** [comment]
Hi @schristm, sorry for the slow reply. What is the value of `@payment.created_at` in a Rails console?

**Comment by schristm** [comment]
`@payment.inspect`
`{ ..., created_at: "2015-07-02 06:30:56", ... }`

`@payment.created_at`
`"Thu, 02 Jul 2015 06:30:56 UTC +00:00"`

**Comment by javan** [comment]
Thanks. See what you get when you run these in your browser's JavaScript console:

```js
> time = new Date(Date.parse("2015-07-02T06:30:56Z"))
Thu Jul 02 2015 02:30:56 GMT-0400 (EDT)

> LocalTime.strftime(time, "%B %e, %Y at %l:%M%P %Z")
"July 2, 2015 at 2:30am EDT"
```

**Comment by schristm** [comment]
I guess it's my browser, but I can't tell why or what I need to do to fix it (and if it'll be a problem our users experience too)
''
> time = new Date(Date.parse("2015-07-02T06:30:56Z"))
< Thu Jul 02 2015 06:30:56 GMT-0400 (EDT)

> LocalTime.strftime(time, "%B %e, %Y at %l:%M%P %Z")
< "July 2, 2015 at 6:30am EDT"
''

**Comment by javan** [comment]
What browser (version, platform, etc)? Maybe it's another script
interfering.

**Comment by briemarie** [comment]
Yes, I am having the same problem. Except the time difference is bigger at 7 hours ahead. 

time = Time.now.in_time_zone(zone) gives me:  2015-07-29 18:07:16 -0700 (which is the correct time and day)
But once I add local_time(time), I get: July 30, 2015 1:08am

I am using Ruby 2.1.5 and Rails 4.1.9.

**Comment by javan** [comment]
@briemarie Are you passing any options to the `local_time` helper? Please copy the resulting element from your browser's console and paste it here.

**Comment by schristm** [comment]
@javan, sorry I meant to respond previously. My browser is Safari 10.10.4, and as far as scripts loaded, it's pretty much just jQuery, Bootstrap, and Google Analytics. I'll try disabling other javascripts and see if the problem goes away.

**Comment by briemarie** [comment]
I was not passing any helpers, but even if I did, such as local_time(@timezone, '%B %e, %Y %l:%M %p'), I was still getting the 7 hours ahead issue.

**Comment by schristm** [comment]
@javan, I disabled all the other javascripts, other than turbolinks, and I've still got the +4 hour difference. I'm sure it's got something to do with my and @briemarie's browser settings, but I'm not sure where to look. I'm also certain other users would encounter the same problem. How is it that rails-timeago doesn't seem to have this issue, and it's also using JavaScript?

**Comment by javan** [comment]
Presumably because of the different approaches to Date parsing: https://github.com/rmm5t/jquery-timeago/blob/master/jquery.timeago.js#L116-L123 and https://github.com/basecamp/local_time/blob/master/app/assets/javascripts/local_time.js.coffee#L6-L17

Can you run `new Date()` in your js console and paste the output please? Here's mine:
```
> new Date()
Thu Jul 30 2015 14:05:41 GMT-0400 (EDT)
```

And run `ls -l /etc/localtime` from the terminal. Here's mine:
```
 $ ls -l /etc/localtime
lrwxr-xr-x  1 root  wheel  35 May  6 10:12 /etc/localtime -> /usr/share/zoneinfo/America/Detroit
```

**Comment by schristm** [comment]
Sure thing.

```
> new Date()
Thu Jul 30 2015 14:25:01 GMT-0400 (EDT)

$ ls -l /etc/localtime
lrwxr-xr-x  1 root  wheel  36 Oct 16  2014 /etc/localtime -> /usr/share/zoneinfo/America/New_York
```

**Comment by briemarie** [comment]
```
new Date()
Thu Jul 30 2015 11:35:33 GMT-0700 (PDT)

~ :> ls -l /etc/localtime
lrwxr-xr-x  1 root  wheel    39B Apr 28  2014 /etc/localtime -> /usr/share/zoneinfo/America/Los_Angeles
```

**Comment by javan** [comment]
@schristm - I've tried several versions of Safari and haven't been able to reproduce. :/

@briemarie - Do you have `local_time` included in your js bundle (e.g `//= require local_time`)? Please try running the following in your browser's js console:

```
> time = new Date(Date.parse("2015-07-02T06:30:56Z"))
Thu Jul 02 2015 02:30:56 GMT-0400 (EDT)

> LocalTime.strftime(time, "%B %e, %Y at %l:%M%P %Z")
"July 2, 2015 at 2:30am EDT"
```

**Comment by schristm** [comment]
@javan, do you have this running this somewhere live I can browse to and test to see if I'm getting the same results?

**Comment by javan** [comment]
Sure. What HTML do you get here? http://codepen.io/javan/pen/GJeNKx

**Comment by schristm** [comment]
July 2, 2015 at 2:30am EDT 

<time data-format="%B %e, %Y at %l:%M%P %Z" data-local="time" datetime="2015-07-02T06:30:56Z" title="July 2, 2015 at 2:30am EDT" data-localized="true">July 2, 2015 at 2:30am EDT</time>

**Comment by javan** [comment]
That's the expected HTML with the correctly localized times so local_time is working as expected. I suspect something else in your app is interfering.

**Comment by schristm** [comment]
@javan, The strange thing is that I already disabled all the other JavaScripts and still get this result. Do you have any advice on how to troubleshoot this deeper?

**Comment by javan** [comment]
@schristm, could you put together a small, fresh rails app that illustrates the issue and put it online somewhere?

**Comment by isabelgm** [comment]
Hi @javan. I'm having a similar issue as well. When I try to create a test for the local_time ago I get a 5 hour time difference. I'm not sure why this is.  I've made sure I have //= require local_time. 

**This is my feature test**
feature "User visits card show page" do
  scenario "after submitting and can see timestamp for completed" do
    user_enrolls_creates_card_and_visits_page
    expect(page).to have_content("Completed on Oct 29, at 6:24 PM")
    expect(page).not_to have_content("Updated at")
  end

  scenario "after updating card and can see timestamp for update" do
    user_enrolls_creates_card_and_visits_page
    expect(page).to have_content("Completed on Oct 29, at 6:24 PM")
    expect(page).to have_content("Updated October 30, 2015 8:24am")
  end
end

def user_enrolls_creates_card_and_visits_page
  @user = create(:user)
  login_as @user
  @project = create(:project)
  @user.enroll(@project)
  @card = create(:card, project: @project, name: "Read book")
  @submission = create(:submission, card: @card, user: @user,
                        created_at: "2015-10-29 18:24:25 -0500",
                        **updated_at:"2015-10-30 08:24:25 -0500")**
  visit submission_path(@submission)

The error I get is: 
Failure/Error: expect(page).to have_content("Updated October 30, 2015 8:24am")
       expected to find text "Updated October 30, 2015 8:24am" in "Toggle navigation Explore New + New Project New Organization Tester Test Profile Edit Profile Log Out Read book Tester Test has completed this card. Completed on Oct 29, at 6:24 PM Updated **October 30, 2015 1:24pm** Back to project. 

Any ideas of what I should be checking?

**Comment by javan** [comment]
@isabelgm can you confirm this is happening in a browser and not just in your test?

**Comment by isabelgm** [comment]
@javan this only happens with the test.

**Comment by javan** [comment]
@isabelgm The conversion to local time happens client side and asynchronously so I wouldn't expect it to pass unless you're testing against a running browser and waiting for `DOMContentLoaded` before scanning the page's content.

**Comment by julianguyen** [comment]
Hi there, I'm running into problems as well. local_time renders the correct date/time on page refreshes but when I am loading things dynamically via Ajax and call local_time, the result is in UTC i.e. https://github.com/julianguyen/ifme/blob/master/app/controllers/application_controller.rb#L269 and https://github.com/julianguyen/ifme/issues/178

**Comment by phongleland** [comment]
Adding the following works for me
`
$(document).on("turbolinks:load", function() {
      LocalTime.run()
  });
`

**Comment by mikevic** [comment]
+1

**Comment by jakehockey10** [comment]
Using this with ActionCable, I had to add `LocalTime.run()` after appending the new message to the chatroom area.  For instance:

```coffeescript
received: (data) ->
    # Called when there's incoming data on the websocket for this channel
    active_chatroom = $("[data-behavior='messages'][data-chatroom-id='#{data.chatroom_id}']")
    if active_chatroom.length > 0

      if document.hidden
        if $('.strike').length == 0
          active_chatroom.append("<div class='strike'><span>Unread Messages</span></div>")

        if Notification.permission == 'granted'
          createNotification(data.username, data.body)

      else
        App.last_read.update(data.chatroom_id)

      # Insert the message
      active_chatroom.append(data.message)
      LocalTime.run()
      active_chatroom.scrollTop(active_chatroom[0].scrollHeight)

    else
      $("[data-behavior='chatroom-link'][data-chatroom-id='#{data.chatroom_id}']").css("font-weight", "bold")
```

**Comment by javan** [comment]
> I had to add `LocalTime.run()` after appending the new message

You can switch over to the 2.0 branch fix that. It observes DOM mutations instead of listening for events. https://github.com/basecamp/local_time/pull/55

**Comment by jakehockey10** [comment]
@javan Thanks for letting me know!  Is there a planned release date for 2.0?

**Comment by jakehockey10** [comment]
@javan Once I got home, I took a look at upgrading it to 2.0 and reminded myself that I had already done this.  I forked `local_time` so that I could merge 2.0 into master and currently have this in my gemfile: `gem 'local_time', github: 'jakehockey10/local_time'`.  But without the addition I mention above, I still have the issue here: https://github.com/jakehockey10/group-chat

**Comment by jakehockey10** [comment]
@javan, I got home to look into this and reminded myself that I had forked the repo with the intention of merging 2.0 into master myself.  I just merged 2.0 into master on my side and can confirm (I doubt you needed further confirmation, but I'll give it to ya anyway ;-) ), and I can say it indeed worked for me.
